src/vm/i386/umthunkstub.S

```diff
@@ -25,8 +26,142 @@ NESTED_ENTRY TheUMEntryPrestub, _TEXT, UnhandledExceptionHandlerUnix
     jmp     eax     // Tail Jmp
 NESTED_END TheUMEntryPrestub, _TEXT
 
-
+//
+// eax: UMEntryThunk*
+//
 NESTED_ENTRY UMThunkStub, _TEXT, UnhandledExceptionHandlerUnix
-    int     3       // implement here
+
+#define UMThunkStub_LOCALVARS               (2*4)   // UMEntryThunk*, Thread*
+#define UMThunkStub_UMENTRYTHUNK_OFFSET     0x04
+#define UMThunkStub_THREAD_OFFSET           0x08
+#define UMThunkStub_INT_ARG_SPILL           (2*4)   // for save ecx, edx
+#define UMThunkStub_FIXEDALLOCSIZE          (UMThunkStub_LOCALVARS+UMThunkStub_INT_ARG_SPILL)
+#define UMThunkStub_INT_ARG_OFFSET          (UMThunkStub_LOCALVARS+4)
+
+// return address                           <-- entry ESP
+// saved ebp                                <-- EBP
+// UMEntryThunk*
+// Thread*
+// save ecx
+// save edx
+// {optional stack args passed to callee}   <-- new esp
+
+    PROLOG_BEG
+    PROLOG_END
+    sub     esp, UMThunkStub_FIXEDALLOCSIZE
+
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET], ecx
+    mov     dword ptr [ebp - UMThunkStub_INT_ARG_OFFSET - 4], edx
+
+    mov     dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET], eax
+
+    call    C_FUNC(GetThread)
+    test    eax, eax
+    jz      LOCAL_LABEL(DoThreadSetup)
+
+LOCAL_LABEL(HaveThread):
+
+    mov     dword ptr [ebp - UMThunkStub_THREAD_OFFSET], eax
+
+    // FailFast if a native callable method invoked via ldftn and calli.
+    cmp     dword ptr [eax + Thread_m_fPreemptiveGCDisabled], 1
+    jz      LOCAL_LABEL(InvalidTransition)
+
+    // disable preemptive GC
+    mov     dword ptr [eax + Thread_m_fPreemptiveGCDisabled], 1
+
+    // catch returning thread here if a GC is in progress
+    PREPARE_EXTERNAL_VAR g_TrapReturningThreads, eax
+    cmp     eax, 0
+    jnz     LOCAL_LABEL(DoTrapReturningThreadsTHROW)
+
+LOCAL_LABEL(InCooperativeMode):
+
+#if _DEBUG
+    mov     eax, dword ptr [ebp - UMThunkStub_THREAD_OFFSET]
+    mov     eax, dword ptr [eax + Thread__m_pDomain]
+    mov     esi, dword ptr [eax + AppDomain__m_dwId]
+
+    mov     eax, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
+    mov     edi, dword ptr [eax + UMEntryThunk__m_dwDomainId]
```

Comment:
@seanshpark I am sorry, I don't know how I could have missed the fact that you are actually using esi adn edi in this function. That means that we need to push them in the prolog and pop them in the epilog to preserve them as they are callee saved registers and we cannot destroy them otherwise we would break our callers that might have something in those registers and expect them being preserved. The same is true for ebx.